pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImplementationBindingImpl.java

```diff
@@ -0,0 +1,348 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.impl;
+
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.Date;
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.pulsar.client.api.Authentication;
+import org.apache.pulsar.client.api.BatcherBuilder;
+import org.apache.pulsar.client.api.ClientBuilder;
+import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.client.api.schema.GenericRecord;
+import org.apache.pulsar.client.api.schema.GenericSchema;
+import org.apache.pulsar.client.api.schema.RecordSchemaBuilder;
+import org.apache.pulsar.client.api.schema.SchemaDefinition;
+import org.apache.pulsar.client.api.schema.SchemaDefinitionBuilder;
+import org.apache.pulsar.client.internal.PulsarClientImplementationBinding;
+import org.apache.pulsar.common.schema.KeyValue;
+import org.apache.pulsar.common.schema.KeyValueEncodingType;
+import org.apache.pulsar.common.schema.SchemaInfo;
+import org.apache.pulsar.common.schema.SchemaInfoWithVersion;
+import org.apache.pulsar.common.schema.SchemaType;
+
+/**
+ * Helper class for class instantiations and it also contains methods to work with schemas.
+ */
+@SuppressWarnings("unchecked")
+public class PulsarClientImplementationBindingImpl implements PulsarClientImplementationBinding {
+
+    public <T> SchemaDefinitionBuilder<T> newSchemaDefinitionBuilder() {
+        return new org.apache.pulsar.client.impl.schema.SchemaDefinitionBuilderImpl();
+    }
+
+    public ClientBuilder newClientBuilder() {
+        return new org.apache.pulsar.client.impl.ClientBuilderImpl();
```

Comment:
Wouldn't this be using the class loader from the current context? How can we be sure it's going to be the same as `PulsarClientImplementationBindingImpl`?

Reply:
See the JVM specs here
https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-5.html
chapter "5.3. Creation and Loading"
look for "If D was defined by a user-defined class loader, then that same user-defined class loader initiates loading of C"

basically if a class "D" (where "D" is not a system class) triggers the loading of a class "C", then the same Classloader that loaded "D" will initiate the loading of "C" 


So in this case the same classloader which loaded `PulsarClientImplementationBindingImpl` will initiate the loading of `ClientBuilderImpl` if the class has not been loaded yet

Reply:
Got it. thanks!